test/e2e/volume_provisioning.go

```diff
@@ -131,6 +132,67 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
 				testDynamicProvisioning(c, claim, "2Gi")
 			}
 		})
+
+		It("should test that deleting a claim before the volume is provisioned deletes the volume. [Volume]", func() {
+			// This case tests for the regressions of a bug fixed by PR #21268
+			// REGRESSION: Deleting the PVC before the PV is provisioned can result in PV
+			// not being deleted.
+			// NOTE:  Checks for regression by iteratively calling .PersistentVolumes.List() and waiting
+			// 10 seconds for 30 iterations (total 2.5 minutes). Will run full time if none are detected.
+
+			var raceAttempts int = 10
+			var suffix string = ns[len(ns)-5:]
+
+			By("Creating a StorageClass")
+			class := newStorageClass("kubernetes.io/gce-pd", suffix)
+			class, err := c.Storage().StorageClasses().Create(class)
+			framework.Logf("Created sc %s", class.Name)
+			Expect(err).NotTo(HaveOccurred())
+			defer c.Storage().StorageClasses().Delete(class.Name, nil)
+
+			// To increase chance of regression, try `raceAttempt` iterations
+			By(fmt.Sprintf("Creating and deleting PersistentVolumeClaims for %d iterations", raceAttempts))
+			claim := newClaim(ns, suffix, false)
+			for i := 0; i < raceAttempts; i++ {
+				tmpClaim, err := c.Core().PersistentVolumeClaims(ns).Create(claim)
+				Expect(err).NotTo(HaveOccurred())
+				time.Sleep(5 * time.Second)
+				err = c.Core().PersistentVolumeClaims(ns).Delete(tmpClaim.Name, nil)
+				Expect(err).NotTo(HaveOccurred())
+			}
+
+			regressedPVs := make([]v1.PersistentVolume, 0)
+			By(fmt.Sprintf("Checking for residual PVs associated with StorageClass %s", class.Name))
+			// Every 10 seconds, check for PVs that have not been Released.  These indicate a regression
+			for i := 0; i < 30 && len(regressedPVs) == 0; i++ {
+				time.Sleep(10 * time.Second)
+				clusterPVs, err := c.Core().PersistentVolumes().List(metav1.ListOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				for _, pv := range clusterPVs.Items {
+					if pv.Annotations[storageutil.StorageClassAnnotation] == class.Name &&
+						pv.Status.Phase != v1.VolumeReleased {
```

Comment:
I thought Jan's comment was to not bother testing Phase. Just `List` all pvs matching the sc annotation and expect the count to be zero eventually.